### ISSUE TITLE
Link some libraries using --whole-archive

### ISSUE THREAD

**Issue by pp5x** [waf-project/waf#2138]
Waf builds static libraries that need to be linked using the `--whole-archive` linker flag. I found that I can do:

~~~.py
self.env.SHLIB_MARKER = ['-Wl,-Bstatic,--whole-archive']`
self.env.SHLIB_MARKER = ['-Wl,-Bdynamic,--no-whole-archive']
~~~

Unfortunately, changing these environment variables has effect on the **entire build**. I do have other static libraries that do not require the `--whole-archive` flag.

### Use case of `--whole-archive`

* Libraries _"plugin"_ which registers on initialization using `__attribute__((constructor))`
* Executable or library which store linked plugins in its global state.

This problem is well described in this Meson issue:  https://github.com/mesonbuild/meson/issues/3046

### Suggestion

A proper way to set dedicated linking options for dependencies (may be it should be generic). For example Meson provides the `link_whole` option when linking against an executable and shared libraries.

The `use` parameter can mix dependencies regardless of there kind (object file, static / shared libraries). I understand it could be cumbersome to have a special case to specify options for a target.

I think a change is required in `ccroot.py` to support this.

#### 1st attempt: add a step before `apply_link` (untouched waf approach)

I tried to add a step before `apply_link` using `@before_method` and `@after_method` decorators to process a new argument called `use_whole`, but it is tricky since `use_whole` applies only for _static library_ dependencies and needs to be processed as well (ie. call `process_use`). This method do not feel comfortable because `process_use` already computes linking flags.

~~~.py
    bld(features='c cprogram whole_archive', target='myprogram', source=[
        # Sources here...
    ],
    use='mylib1 mylib2', # Libraries that do no need --whole-archive
    use_whole='mywholelib', # Libraries that require --whole-archive
    lib=['pthread', 'dl', 'm']) # External libraries
~~~

- - -
tl;dr: Have a way to set link options on given dependencies without side effect.

Do you have any pointer to solve this issue? I could look into to it and provide a patch for waf if needed. Thank you!

**Comment by ita1024** [comment]
The following should be sufficient for the above:
```python
tg = bld(features='cxx cxxprogram', ...)
tg.env.SHLIB_MARKER = ['-Wl,-Bdynamic,--no-whole-archive']
```

Now, if you need something more fine-grained, add a method after `process_use` and then move the desired library flags from `self.env.LIB/self.env.STLIB` to `self.env.LDFLAGS/self.env.LINKFLAGS` (or assign new copies of those lists to `self.link_task.env.XYX`). You can even re-order the libraries entirely if you need to.

Adding an extension for waf is unlikely as there is a variety of other complex or conflicting cases (`--start-group...`, order of libraries, object files...). Implement such functionality in your own project.
